### PR TITLE
Add fast permissions endpoint with roaring bitmaps

### PR DESCRIPTION
**Pull request by kzh**
This PR contains a complete and functioning implementation of RFC 45 ACL endpoints via roaring bitmaps.

TODO:
- [ ] ~RFC for technical specification (rest api, etc)~
- [ ] Flush out the PR description
- [ ] Docs

Endpoints:
```
/permissions/users?repository=PROJECT_1%2Frep_1&permission=read
/permissions/repositories?user=user&permission=admin
```

Some numbers:
`https://bitbucket.sgdev.org/rest/sourcegraph-admin/1.0/permissions/repositories?user=milton&permission=read` returns 10k repository ids in 700ms.

This is a part of sourcegraph/sourcegraph#6091.

### CODE REVIEW

**Comment by tsenart** [comment]
Before I review this code, I'd like you to load test these new endpoints in bitbucket.sgdev.org with enough repos to mimic some of our largest customers that use Bitbucket Server (i.e. ~ 32k repos).

How does the p99 latency of these endpoints look at 5 req/s, 10 req/s and 50 req/s?

This benchmarking script may be of help: https://github.com/tsenart/vegeta/pull/451

File: pom.xml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
```
Comment (keegancsmith) [inline pom.xml:3]: is the reformatting in this file due to your editor doing it? I assume this doesn't change anything (specifically the description section has some reflowing)

File: pom.xml [hunk 2]
```diff
@@ -46,6 +49,12 @@
     </dependencyManagement>
 
     <dependencies>
+        <dependency>
+            <groupId>org.roaringbitmap</groupId>
+            <artifactId>RoaringBitmap</artifactId>
+            <version>0.6.66</version>
```
Comment (keegancsmith) [inline pom.xml:55]: latest version is `0.8.11`. Any reason you picked this version?

File: src/main/java/com/sourcegraph/permission/PermissionRouter.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        PermissionRouter.repositoryService = repositoryService;
+        PermissionRouter.userManager = userManager;
+        PermissionRouter.userService = userService;
+        PermissionRouter.securityService = securityService;
+    }
+
+    // The getUsersWithRepositoryPermission endpoint returns a roaring bitmap containing the IDs of all the users
+    // that is granted the specified permission to the specified repository.
+    //
+    // Ex. /permissions/users?repository=PROJECT_1/rep_1&permission=read
```
Comment (keegancsmith) [inline src/main/java/com/sourcegraph/permission/PermissionRouter.java]: minor: Why not use java doc? I think you need to do `/**` style block for java doc to pick up on it.

File: src/main/java/com/sourcegraph/permission/PermissionRouter.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        }
+
+        Repository repository = getRepository(repo);
+        if (repository == null) {
+            return Response.status(Response.Status.NOT_FOUND).entity("No such repository: " + repository).build();
+        }
+
+        Permission permission = getRepositoryPermission(perm);
+        if (permission == null) {
+            return Response.status(Status.UNPROCESSABLE_ENTITY).build();
```
Comment (keegancsmith) [inline src/main/java/com/sourcegraph/permission/PermissionRouter.java]: minor: Can you include a message explaining that the permission is invalid and include the list of valid permissions?

File: src/main/java/com/sourcegraph/permission/PermissionRouter.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        do {
+            Page<ApplicationUser> page = userService.search(search, pageRequest);
+            for (ApplicationUser user : page.getValues()) {
+                bitmap.add(user.getId());
+            }
+            pageRequest = page.getNextPageRequest();
+        } while (pageRequest != null);
+
+        byte[] backing = serialize(bitmap);
+        return backing != null ? Response.ok(backing).build() : Response.serverError().build();
```
Comment (keegancsmith) [inline src/main/java/com/sourcegraph/permission/PermissionRouter.java]: - can you include in the response why it failed for serverError?
- Will the response include a header for the size of the response?
- Can you include a content-type header in the response? If you are unsure you can use `application/octet-stream`. But I have another comment about some basic content negotiation so we can get json back for debugging purposes. In that case you will likely need to make up a mime type for roaring bitmaps.

File: src/main/java/com/sourcegraph/permission/PermissionRouter.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            return Response.status(Response.Status.NOT_FOUND).entity("No such user: " + user).build();
+        }
+
+        Permission permission = getRepositoryPermission(perm);
+        if (permission == null) {
+            return Response.status(Status.UNPROCESSABLE_ENTITY).build();
+        }
+
+        EscalatedSecurityContext context = securityService.impersonating(user, "Repository Search");
+        RoaringBitmap bitmap = context.call(() -> {
```
Comment (keegancsmith) [inline src/main/java/com/sourcegraph/permission/PermissionRouter.java:116]: neat, java is much nicer than I remember

**Review by keegancsmith — CHANGES_REQUESTED** [review]
This is great stuff, looking forward to seeing how much faster this is than the previously naive approach we used. Requesting changes cause I wanna take an other look after addressing my feedback.

- Please add some basic unit tests. Doesn't need to be in-depth.
- Some very simple content negotation so we can get back either JSON or roaring bitmap would be ideal. This will greatly simplify debugging when implementing this on the client side. You can potentially do that as a follow-up PR.
- There may be potential for using roaring bitmaps more efficiently. I think there are APIs around batch insertion of numbers which amortize some of the fixed costs for inserting a number.

File: pom.xml (hunk 1 above)
Comment (kzh) [reply pom.xml:3]: Yep, this was from IntelliJ's formatting.

File: pom.xml (hunk 2 above)
Comment (kzh) [reply pom.xml:55]: Newer versions are incompatible with the Java version that Bitbucket's web framework (Jersey) supports. It's a strange case of version compatibility, but anything past 0.7 crashes with an obscure exception 🤔

https://github.com/RoaringBitmap/RoaringBitmap#prerequisites
```
Version 0.7.x requires JDK 8 or better
Version 0.6.x requires JDK 7 or better
Version 0.5.x requires JDK 6 or better
```

File: src/main/java/com/sourcegraph/permission/PermissionRouter.java (hunk 5 above)
Comment (kzh) [reply src/main/java/com/sourcegraph/permission/PermissionRouter.java]: > Will the response include a header for the size of the response?

The response includes the header `content-length` (bytes). I also added a debug header for cardinality of the bitmap (`X-Debug-Count`).

>  In that case you will likely need to make up a mime type for roaring bitmaps.

Got it! I'll use `application/roaring-bitmap`.

File: src/main/java/com/sourcegraph/permission/PermissionRouter.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+            return Response.status(Status.UNPROCESSABLE_ENTITY).build();
+        }
+
+        RoaringBitmap bitmap = new RoaringBitmap();
+
+        UserSearchRequest.Builder builder = new UserSearchRequest.Builder();
+        builder.repositoryPermission(repository, permission);
+        UserSearchRequest search = builder.build();
+
+        PageRequest pageRequest = new PageRequestImpl(0, 5000);
```
Comment (unknwon) [inline src/main/java/com/sourcegraph/permission/PermissionRouter.java]: How "5000" gets chosen?

**Review by unknwon — APPROVED** [review]

File: src/main/java/com/sourcegraph/permission/PermissionRouter.java (hunk 7 above)
Comment (kzh) [reply src/main/java/com/sourcegraph/permission/PermissionRouter.java]: It's an arbitrarily chosen value. I'm not exactly sure how it impacts memory as of yet. 

These were the endpoint response times for 10,000 repository IDs.
`100` page size -> 3.7s
`5000` page size -> 0.7s

**Comment by tsenart** [comment]
@kzh: Can you please write a summary of the status of this work? How did the load testing go? Please share those results too.

File: src/main/java/com/sourcegraph/permission/PermissionRouter.java (hunk 7 above)
Comment (unknwon) [reply src/main/java/com/sourcegraph/permission/PermissionRouter.java]: Gotcha, would you mind put what you replied as a block of comments in the code?

**Comment by kzh** [comment]
Last week, I did a bit of load testing that resulted in borking [bitbucket.sgdev.org](bitbucket.sgdev.org). After doing some investigation, I found out that the endpoints don't perform well concurrently. This appears to be due to Jersey 1.0.4 (Bitbucket's REST framework, released 9 years ago) not supporting non blocking/asynchronous request handlers. I'm currently looking into a workaround. I am somewhat confident that this will be possible.

I am prioritizing this over the other tasks.

The endpoint performs decently at one request per second. There is a lot of room for optimizing this latency.
```
$ echo "GET https://milton:<password>@bitbucket.sgdev.org/rest/sourcegraph-admin/1.0/permissions/repositories?user=milton&permission=read" | vegeta attack -duration=60s -rate=1/s | tee results.bin | vegeta report --every=1s

Requests      [total, rate, throughput]  60, 1.02, 1.00
Duration      [total, attack, wait]      59.751935948s, 59.000064102s, 751.871846ms
Latencies     [mean, 50, 95, 99, max]    788.082419ms, 771.964092ms, 946.331521ms, 1.022288631s, 1.024401939s
Bytes In      [total, mean]              492480, 8208.00
Bytes Out     [total, mean]              0, 0.00
Success       [ratio]                    100.00%
Status Codes  [code:count]               200:60
Error Set:
```

**Comment by keegancsmith** [comment]
> I am prioritizing this over the other tasks.

That sounds like it can be follow-up work if we validate this is faster than our older method of fetching all repos for a user. We should focus on merging this in / validating it is a correct approach.

> The endpoint performs decently at one request per second. There is a lot of room for optimizing this latency.

That is great! Can you also do a load test for our older method (listing repos) so we can compare the performance? If the perf of this is better it validates our approach and we can merge in.

**Comment by kzh** [comment]
> Can you also do a load test for our older method (listing repos) so we can compare the performance? If the perf of this is better it validates our approach and we can merge in.

Yes, will do! I spoke with Tomás very briefly on this. Some very rough numbers (based on 10k repository result): the latency on the older method is several seconds (up to 10s) versus 800ms with this newer approach, so there is a high probability that this is a somewhat substantial improvement. I will do a more intensive review/load test to verify this.

I also did some manual profiling. The performance bottleneck comes from the Bitbucket Search API. The creation/population of the roaring bitmap portion is negligible on the performance. So there isn't much to optimize. I can experiment with using another part of the Bitbucket SDK to see if it's faster, but I don't think this is that big of a concern at the moment.

**Update on concurrency**: Upon a lot of research, I concluded that Jersey 1.0.2 (the outdated web framework that Atlasssian Bitbucket uses) does not support asynchronous request handlers. This means that only one request can be processed at a time. This is concerning since this endpoint will be called frequently for refreshing a permission cache on the Sourcegraph instance per user at a specified TTL rate.

There are two workarounds that I can think of for addressing this problem. They both require additional overhead. At the moment, we send a single http request to use this endpoint. I propose that we change it to either:
* Polling - we send an initial request to start the job (searching, building roaring bitmap). This gets processed on a separate thread. Then we send subsequent requests to check on the job for completion.
* Push - this approach is similar to how a webhook works. We send a request to initiate the task. This gets processed on a separate thread. When the task is finished, the plugin sends a request back to the Sourcegraph instance with the data.

**Comment by tsenart** [comment]
> So there isn't much to optimize. I can experiment with using another part of the Bitbucket SDK to see if it's faster, but I don't think this is that big of a concern at the moment.

That assumes that the performance of that code path stays below a given threshold under concurrency. We don't know that yet, since we haven't figured out how to handle HTTP requests concurrently.

> Polling

How many days of development would you say validating this approach would take?

> Push

This could work, but it'd take substantial effort and complexity to integrate on Sourcegraph's side, rather than the straightforward change it would be otherwise.

I am surprised that the Bitbucket Server API can't handle requests concurrently. Is this true for it's normal endpoints or only for these extension provided endpoints?

**Comment by kzh** [comment]
> Polling
>> How many days of development would you say validating this approach would take?

It's a small change so it would take a couple of hours to two days.

> I am surprised that the Bitbucket Server API can't handle requests concurrently. Is this true for it's normal endpoints or only for these extension provided endpoints?

Yeah, I find this bizarre too. I'm not sure if I overlooked something. I think this only applies to endpoints from plugins.

The Bitbucket REST plugin API is on an outdated version of Jersey (1.0.2, 2010). Jersey 2.X (>2014) introduces asynchronous requests via `@ManagedAsync` and `@Suspended AsyncResponse`.

---
https://community.developer.atlassian.com/t/jax-rs-2-0-support/13263
https://community.developer.atlassian.com/t/jax-rs-2-0-support-in-atlassian-rest-module/16126/5
https://developer.atlassian.com/server/framework/atlassian-sdk/rest-plugin-module/
> The REST module is based on version 1.0.2 of Jersey.

https://bitbucket.org/atlassian/atlassian-rest/src/df33341de7d9df4a403c0f6aba96810d1f356238/pom.xml#lines-57

**Comment by mrnugget** [comment]
(I've been following the discussion and I think there's a slight chance of "talking about different things" here that can be cleared up)

> > I am surprised that the Bitbucket Server API can't handle requests concurrently. Is this true for it's normal endpoints or only for these extension provided endpoints?
> 
> Yeah, I find this bizarre too. I'm not sure if I overlooked something. I think this only applies to endpoints from plugins.
> 
> The Bitbucket REST plugin API is on an outdated version of Jersey (1.0.2, 2010). Jersey 2.X (>2014) introduces asynchronous requests via `@ManagedAsync` and `@Suspended AsyncResponse`.

Are concurrent requests the same as asynchronous requests?

Concurrent requests, to me, would mean that a web server can multiple requests at the same time.

So, here's my question in regards to this:

> **Update on concurrency**: Upon a lot of research, I concluded that Jersey 1.0.2 (the outdated web framework that Atlasssian Bitbucket uses) does not support asynchronous request handlers. This means that only one request can be processed at a time. This is concerning since this endpoint will be called frequently for refreshing a permission cache on the Sourcegraph instance per user at a specified TTL rate.

Why would asynchronous requests be necessary so that more than one request "can be processed at a time"?

I'm asking because usually webservers spawn multiple processes or threads to handle concurrent requests, or they use an event loop, etc.

**Comment by tsenart** [comment]
@kzh: I know you've been speaking with @keegancsmith about this PR and how to proceed. Please provide an action plan with next steps and a timeline in a reply to this comment.

**Comment by kzh** [comment]
`<WIP comment>`

@mrnugget 
Ah, sorry! I mistakenly used `asynchronous` and `concurrent` interchangeably. I misread what was happening. It is as you described! Jersey does support concurrent calls via spawning threads. The culprit was the expensive performance bottleneck that was hogging all the threads in the pool (thus blocking new requests). I elaborate a bit on this below.

Asynchronous support can help with this by separating request handling and worker operations into individual thread pools, so one won't be able to overwhelm the other (https://dzone.com/articles/jax-rs-20-asynchronous-server-and-client).

**Update**: So I did some more debugging alongside the polling implementation. The root of the concurrency problem can be narrowed down to Bitbucket's repository search API.
```
repositoryService.search(search, pageRequest)
```
This API call alone accounts for 99% of the time. During one of the load tests, it took 85s. I'm assuming it either has a limited thread pool or doesn't support concurrent calls (blocks). I haven't verified this yet. This method normally takes 500ms.

Someone posted about this in 2016.
https://community.atlassian.com/t5/Answers-Developer-Questions/bitbucket-performance-repositoryservice-userservice-paging-api/qaq-p/501841

**Comparison**:
Old: https://bitbucket.sgdev.org/rest/api/1.0/repos?limit=1000
New: https://bitbucket.sgdev.org/rest/sourcegraph-admin/1.0/permissions/repositories?user=milton&permission=read

The old (current) endpoint is slightly faster (~500ms) but is limited to 1000 results per request/page. So for 10k repositories, it would require ten requests (total taking ~4.5 seconds). The advantage of the newer endpoint is that all repositories are included in one request and it takes around 750ms.

However, a big issue that both of these endpoints suffer from is that they both face similar significant performance deterioration as QPS increases over an extended period. They don't scale well. I believe they both call onto the same underlying API `RepositoryService.search` which is where the bottleneck occurs.

**Action plan** (a few hours):  
- [ ] Merge this PR
- [ ] Implement bitbucket server authz provider `PermissionsUpdateFunc` to fetch from this endpoint (if plugin is installed) behind a feature flag: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/frontend/internal/authz/bitbucketserver/provider.go#L134
https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/frontend/internal/authz/bitbucketserver/store.go#L144
- [ ] Spin up a temporary k8s sourcegraph instance (synced with bitbucket.sgdev.org) for performance testing.
- [ ] See if batching would be the appropriate next step. If so, implement.

**Comment by keegancsmith** [comment]
> Update: So I did some more debugging alongside the polling implementation. The root of the concurrency problem can be narrowed down to Bitbucket's repository search API.

Cool, glad you got a head start on that. But can the result of that work be a follow-up PR? It would be great to get this PR merged in so that the discussion of future work is more targetted / incremental.

> However, a big issue that both of these endpoints suffer from is that they both face similar significant performance deterioration as QPS increases over an extended period.

Ok so even with concurrency, this endpoint is probably better than our current method? Is that correct?

> Action plan (a few hours):

Can the first item be merging this PR?

> Implement bitbucket server authz provider PermissionsUpdateFunc to fetch from this endpoint (if plugin is installed) behind a feature flag: 

:+1:

>  Spin up a temporary k8s sourcegraph instance (synced with bitbucket.sgdev.org) for performance testing.

This sounds like a lot of work, and you have already done a lot of validation of this. I think the better approach is to integrate into sourcegraph behind a feature flag. Then when we do validation we can just do it on an existing test cluster we have that is at scale (with the feature flag on).

**Pull request merged**